### ISSUE TITLE
ability to customize data-registration rules/logic

### ISSUE THREAD

**Issue by GlenDC** [rivine/rivine#126]
Currently all transactions which wish to register arbitrary data have to prefix this data with the Prefix `NonSia`, this prefix should probably be customizable per blockchain implementation.

The transaction fee is currently also hardcoded, while it should probably also be customizable for each blockchain implementation, even though it should have sane defaults.

> The transaction fee for the tfchain is for example currently desired to be set at `0.1 TFT`, but those requirements can change.

An open question is also how complex we should allow those transaction fees (related to registration data should be). Should a transaction which wishes to register arbitrary data cost more? If so, should its cost increase together with the size of that arbitrary data.

**Comment by LeeSmet** [comment]
While we are working on this, we should probably also think about custom transaction rules in general. In general, I feel it might be best to allow people to supply a custom function with a predefined signature (something like `func (Transaction) Currency`), taking a transaction as input and calculating a required minimum transaction fee which is then returned. Then everyone could make these as complex as they want. The same could apply to other validation rules

**Comment by GlenDC** [comment]
Does sounds like it would give a lot of flexibility indeed. Could you elaborate on what other validation rules you had in mind? So we're all on the same page.

**Comment by LeeSmet** [comment]
I was referring to this mainly with the idea of the authorized addresses in mind. But coming to think of it, that might be more difficult because you need to know if an address is authorized, which is beyond the context of just a transaction. Though we could also pass a reference to the consensus set as well as this would then make it easy to get a reference to the current height and or other things to make it easy to create soft forks. Disregarding that it feels useful even if its just in case that someone want to create some crazy logic e.g. I want every transaction to create an even number of outputs I guess

**Comment by LeeSmet** [comment]
Could also do this to establish something like a minimum transaction amount: You must s(p)end at leat 0.01C fore example

**Comment by GlenDC** [comment]
Yeah it's true, but again, as Rivine is supposed to be a protocol to implement a blockchain, we should allow flexibility where it is possible. These rules, while having sane defaults (e.g. a minimum transaction amount), should be able to change, when required.

**Comment by GlenDC** [comment]
As part of #165, the NonSia prefix has been removed and is no longer needed.

The only thing left to do for this issue would be provide the option to customize the transaction fees based on stuff like data size. But that is not important for now, so moving it to a later milestone.
